feat(bazel): simplify ng_package by dropping esm5 and fesm5

IgorMinar opened the pull request:
esm5 and fesm5 are no longer needed and have been deprecated in the past.

https://v9.angular.io/guide/deprecations#esm5-and-fesm5-code-formats-in-angular-npm-packages

This commit modifies ng_package to no longer distribute these two formats in npm packages
built by ng_package (e.g. @angular/core).

This commit intentionally doesn't fully clean up the ng_package rule to remove all traces of esm5 and fems5
build artifacts as that is a bigger cleanup and currently we are narrowing down the scope of this change
to the MVP needed for v10, which in this case is 'do not put esm5 and fesm5' into the npm packages.

More cleanup to follow: https://angular-team.atlassian.net/browse/FW-2143

IgorMinar committed:
feat(bazel): simplify ng_package by dropping esm5 and fesm5

esm5 and fesm5 are no longer needed and have been deprecated in the past.

https://v9.angular.io/guide/deprecations#esm5-and-fesm5-code-formats-in-angular-npm-packages

This commit modifies ng_package to no longer distribute these two formats in npm packages
built by ng_package (e.g. @angular/core).

This commit intentionally doesn't fully clean up the ng_package rule to remove all traces of esm5 and fems5
build artifacts as that is a bigger cleanup and currently we are narrowing down the scope of this change
to the MVP needed for v10, which in this case is 'do not put esm5 and fesm5' into the npm packages.

More cleanup to follow: https://angular-team.atlassian.net/browse/FW-2143

BREAKING CHANGE: esm5 and fesm5 format is no longer distributed in
Angular's npm packages e.g. @angular/core

If you are not using Angular CLI to build your application or library,
and you need to be able to build es5 artifacts, then you will need to
downlevel the distributed Angular code to es5 on your own.

Angular CLI will automatically downlevel the code to es5 if differential
loading is enabled in the Angular project, so no action is required from
Angular CLI users.

IgorMinar committed:
build: simplify package.jsons for all of our packages

We can remove all of the entry point resolution configuration from the package.json
in our source code as ng_package rule adds the properties automatically and correctly
configures them.

This change simplifies our code base but doesn't have any impact on the package.json
in the distributed npm_packages.

mary-poppins commented:
You can preview 1a31377 at https://pr36944-1a31377.ngbuilds.io/.

mary-poppins commented:
You can preview b16977a at https://pr36944-b16977a.ngbuilds.io/.

IgorMinar commented:
some of the integration tests are still failing, but the core part of the change is done and ready for the review.

mary-poppins commented:
You can preview c5b9aa6 at https://pr36944-c5b9aa6.ngbuilds.io/.

IgorMinar commented:
@devversion or @alan-agius4 if you have free cycles could you please help me resolve the last remaining test failures? I suspect that most of them are just fs path updates (fesm5->fesm2015) and golden file updates.

if you can't I'll keep on working on this file tomorrow morning.

mary-poppins commented:
You can preview 81c15e5 at https://pr36944-81c15e5.ngbuilds.io/.

devversion committed:
fix(bazel): ng_package rule should update "package.json" of ts_library targets

In the past we added support for `ts_library` to `ng_package`. For those
targets we never can determine the "index" file. Unlike `ng_module`,
there is no provider data for flat module bundles, so the `ng_package`
rule assumes that the index file is simply called `index`.

This works as expected, but we also added logic in the past that doesn't
allow `ng_package` to add format properties (e.g. `main`, `module`) to a
`package.json` if a package json is handwritten for such a `ts_library` target.

This has been done that way because we assumed that such `package.json` files
might want to set format properties explicitly to different paths due to a
faulty "index" guess.

We want to change this behavior as most of the time a `package.json`
file already exists with just the module name. In those cases, the
packager should still set the format properties. We should only warn
and skip automatic insertion of the format properties if such a
`package.json` explicitly sets format properties.

devversion commented:
@googlebot I consent.

mary-poppins commented:
You can preview 6f4880f at https://pr36944-6f4880f.ngbuilds.io/.

alan-agius4 commented:
@IgorMinar, @devversion offered to take this up and try to make it green.

IgorMinar commented on packages/bazel/src/ng_package/packager.ts:
```diff
@@ -342,6 +348,19 @@ function main(args: string[]): number {
+      console.error(
+          '    Skipping automatic insertion of format properties as explicit ' +
+          'format properties are set.');
+      console.error('    Ignore this warning if explicit properties are set intentionally.');
```
this DX is not great. Any time we print a warning that says "ignore this warning if ..." we failed. Could we instead say "add `customFormatResolutions = True` attribute to the target if you really want to provide custom format resolution in your npm package" - needs a bit more word-smithing :-)

devversion replied:
Agreed 👍 

We will have a follow-up for this as we have a similar warning already and it might make sense to think about these warnings in a combined sense. i.e. having the flag disable both of these warnings.

IgorMinar reviewed (commented):
otherwise @devversion's addition looks good to me.

devversion committed:
feat(ngcc): support for new APF where `module` points to esm2015 output

As of version 10, libraries following the APF will no longer contain
ESM5 output. Hence, tests in ngcc need to be updated as they currently
rely on the release output of `@angular/core`.

Additionally, we'd need to support in ngcc that the `module`
property of entry-points no longer necessarily refers to
`esm5` output, but instead can also target `esm2015`.

We currently achieve this by checking the path the `module`
property points to. We can do this because as per APF, the
folder name is known for the esm2015 output. Long-term for
more coverage, we want to sniff the format by looking for
known ES2015 constructs in the file `module` refers to.

devversion committed:
build: update side-effects integration test to not expect esm5 output

As mentioned in previous commits, as of v10 the release output
does no longer contain ESM5 output due to an update to the APF.

This means that the side-effect integration test needs to be
updated as it currently expects/tests esm5 output.

mary-poppins commented:
You can preview 26571e1 at https://pr36944-26571e1.ngbuilds.io/.

mary-poppins commented:
You can preview f30efea at https://pr36944-f30efea.ngbuilds.io/.

mary-poppins commented:
You can preview d19afda at https://pr36944-d19afda.ngbuilds.io/.

devversion committed:
test: disable ng_package tests in ivy test jobs

Since we no longer hardcode the `package.json` for
entry-points, a bug has appeared for `ng_package` in Ivy.

The `package.json` files are populated incorrectly with Ivy
as the flat module bundle name is not propagated from `ng_module`
to the `ng_package` rule. The rule then guesses the index file
to `index.js` and does not respect the flat module bundle shim.

mary-poppins commented:
You can preview 37f6a96 at https://pr36944-37f6a96.ngbuilds.io/.

IgorMinar commented on packages/bazel/test/ng_package/BUILD.bazel:
```diff
@@ -45,6 +49,10 @@ jasmine_node_test(
+    # The "ng_package" rule currently does not properly handle flat module
+    # bundles which the common package tests rely on. We temporarily disable
+    # this test until we fix `ng_package` for Ivy. Tracked with: FW-2144.
+    tags = ["fixme-ivy-aot"],
```
lgtm! thank

mary-poppins commented:
You can preview 3f8cf12 at https://pr36944-3f8cf12.ngbuilds.io/.

IgorMinar committed:
test(docs-infra): temporarily disable the 'upgrade-phonecat-2-hybrid' docs example test

This is due to issue described in: https://github.com/angular/angular/issues/36957

This commit works around the problem rather than trying to solve it in order to unblock the big APF v10 change.

mary-poppins commented:
You can preview c353e14 at https://pr36944-c353e14.ngbuilds.io/.

petebacondarwin commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -89,11 +116,11 @@ runInEachFileSystem(() => {
       expect(() => mainNgcc({
                basePath: '/dist',
-               propertiesToConsider: ['es2015', 'fesm5', 'module'],
+               propertiesToConsider: ['es2015', 'module'],
```
why this change?

IgorMinar replied:
@devversion can you please answer?

I'm guessing that Paul saw that we removed fesm5 from everywhere and try to remove it from here. @petebacondarwin now that you mentioned this, it's likely that this change is a mistake and the line should be rolled back.

IgorMinar replied:
... because existing libraries on npm still use `fesm5` and ngcc might want to consider them...

devversion replied:
Yep, good catch. I think since we don't rely on any of the actual release packages here, this change is unnecessary. Reverting!

devversion replied:
Just to be clear: I did not remove `esm5` everywhere. This particular was just an oversight. I kept dedicated esm5 test as we could process libraries shipping with `<= APF 10`.

Primarily to ensure we still handle ES5 properly, I added a build target that downlevels the core `fesm2015` bundle to `fesm5` using Babel. That way we still get "integration test" coverage for ES5.

petebacondarwin commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -280,7 +311,7 @@ runInEachFileSystem(() => {
       });
 
-      compileIntoFlatEs5Package('test-package', {
+      compileIntoFlatEs2015Package('test-package', {
```
Why this change?
Could you not just change line 347 to be the following?

```
propertiesToConsider: ['esm5', 'esm2015'],
```

devversion replied:
I was thinking that in APF primarily we ship ES2015 code now, and this test doesn't mention any es5 specifics, so I changed it to the flat `es2015` package creation.

Happy to revert if you think we should.

petebacondarwin replied:
I think it is safe to leave it as-is.

petebacondarwin commented on packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:
```diff
@@ -389,7 +420,7 @@ runInEachFileSystem(() => {
     // This is necessary to ensure XPipeDef.fac is defined when delegated from injectable def
     it('should always generate factory def (fac) before injectable def (prov)', () => {
-      compileIntoFlatEs5Package('test-package', {
+      compileIntoFlatEs2015Package('test-package', {
```
Similarly here.

devversion replied:
Same response as above.

petebacondarwin reviewed (approved):
LGTM - couple of minor nits.

Reviewed for fw-ngcc

IgorMinar committed:
docs: remove the note about @angular/http from the deprecations guide

We stopped publishing @angular/http to npm a while back (v7?), so from the perspective of our
customer @angular/http has been removed a long time ago and mentioninig it now that we actually
removed the code from our repo is just confusing.

kara reviewed (approved):
LGTM

Reviewed-for: global-approvers

devversion committed:
fixup! feat(ngcc): support for new APF where `module` points to esm2015 output

Revert unnecessary test change in ngcc spec

mary-poppins commented:
You can preview 13e36fd at https://pr36944-13e36fd.ngbuilds.io/.

devversion reviewed (approved):
LGTM for the changes I did not author.

alan-agius4 commented:
Super excited for this, farewell ES5 👋
